jax2onnx/converter/ir_optimizations.py

```diff
@@ -0,0 +1,1741 @@
+    """
+    ins = _node_inputs(node)
+    for iv in ins:
+        if obj is not None and iv is obj:
+            return True
+        if name:
+            ivn = _v_name(iv)
+            if ivn == name:
+                return True
+            # If inputs are plain strings in this build
+            if isinstance(iv, str) and iv == name:
+                return True
+    return False
+
+
+def _count_consumers(nodes: List[object], name: Optional[str], obj) -> int:
+    """
+    Count how many nodes consume the given value (by name or object).
+    """
+    c = 0
+    for n in nodes:
+        if _has_input_name_or_obj(n, name, obj):
+            c += 1
+    return c
+
+
+def _find_next_consumer_idx(
+    nodes: List[object], start_idx: int, name: Optional[str], obj
+) -> Optional[int]:
+    """
+    Find the index of the next node (after start_idx) that consumes the given
+    value (by name or object). Return None if not found.
+    """
+    for i in range(start_idx + 1, len(nodes)):
+        if _has_input_name_or_obj(nodes[i], name, obj):
+            return i
+    return None
+
+
+# ---------------- IR helpers ----------------
+
+
+def _v_name(v: ir.Value | None) -> Optional[str]:
+    if v is None:
+        return None
+    if isinstance(v, str):
+        return v or None
+    nm = getattr(v, "name", None)
+    return nm if isinstance(nm, str) and nm != "" else None
+
+
+def _value_dtype_code(val: Optional[ir.Value]) -> Optional[int]:
+    if val is None or isinstance(val, str):
+        return None
+    # Value may expose dtype directly or via .type
+    dt = getattr(val, "dtype", None)
+    if isinstance(dt, ir.DataType):
+        return int(dt.value)
+    if isinstance(dt, (int, np.integer)):
+        return int(dt)
+    ty = getattr(val, "type", None)
+    if ty is not None:
+        tdt = getattr(ty, "dtype", None)
+        if isinstance(tdt, ir.DataType):
+            return int(tdt.value)
+        if isinstance(tdt, (int, np.integer)):
+            return int(tdt)
+        elem = getattr(ty, "elem_type", None)
+        if isinstance(elem, ir.DataType):
+            return int(elem.value)
+        if isinstance(elem, (int, np.integer)):
+            return int(elem)
+    return None
+
+
+def _node_outputs(n) -> List["ir.Value"]:
+    outs = getattr(n, "outputs", None)
+    if outs is None:
+        outs = getattr(n, "output", [])  # defensive
+    return list(outs)
+
+
+def _node_output(n) -> Optional["ir.Value"]:
+    outs = _node_outputs(n)
+    return outs[0] if outs else None
+
+
+def _node_inputs(n) -> List["ir.Value"]:
+    ins = getattr(n, "inputs", None)
+    if ins is None:
+        ins = getattr(n, "input", [])  # defensive
+    return list(ins)
+
+
+def _set_node_inputs(n, new_ins: List["ir.Value"]) -> None:
+    """
+    Robustly write a node's inputs across onnx_ir variants that use either
+    `.inputs` (Value objects) or `.input` (Value objects or names).
+    """
```

Comment:
This seems redundent. There is always node.inputs defined (assuming you are not dealing proto objects anymore)